File: javascript/hamming/hamming.js
var Hamming = function() {};

Hamming.prototype.compute = function(a_dna, b_dna) {
  if(a_dna.length != b_dna.length) {
    throw new Error('DNA strands must be of equal length.');
  }
  
  return this.calculate_difference(a_dna, b_dna);
}

Hamming.prototype.calculate_difference = function(a_dna, b_dna) {
  var count = 0;
  for(var i = 0; i < a_dna.length; i++) {
    if(a_dna[i] != b_dna[i]) {
      count++;
    }
  }

  return count;
}

module.exports = Hamming;

File: elixir/hamming/hamming.exs
defmodule Hamming do
  @doc """
  Returns number of differences between two strands of DNA, known as the Hamming Distance.

  ## Examples

  iex> Hamming.hamming_distance('AAGTCATA', 'TAGCGATC')
  {:ok, 4}
  """
  @spec hamming_distance([char], [char]) :: {:ok, non_neg_integer} | {:error, String.t()}
  def hamming_distance(strand1, strand2) when length(strand1) != length(strand2) do
    {:error, "Lists must be the same length"}
  end

  def hamming_distance(strand1, strand2) do
    Enum.zip(strand1, strand2)
    |> Enum.count(&(elem(&1, 0) != elem(&1, 1)))
    |> build_ok_response()
  end

  defp build_ok_response(count) do
    {:ok, count}
  end
end
